### ISSUE TITLE
Graph creation functions

### ISSUE THREAD

**Issue by tomkom** [cwida/duckpgq-extension#57]
-- Preamble: Just stumbled upon the extension, very excited, thank you!

Functionality suggestion ( maybe already in pipeline): 

Explicit graph data are not that common and often have to be created by recomputing the topology. This is common in the spatial context ( street lanes), where graphs are computed by analyzing topological relationships between member geometries. 

The result is a data structure -- the edge table. See for example the behaviour of [pg_routing] (https://docs.pgrouting.org/latest/en/pgr_createTopology.html), potentially with an additional [vertices table](https://docs.pgrouting.org/latest/en/pgr_createVerticesTable.html).

It would be awesome if it was possible to create the input structures in DUCK/PGQ, instead of having to craft them separately.

**Comment by Dtenwolde** [comment]
Hi Tom, 

Thank you :) I'll start working more on the extension again in the near future (wrapping up an internship right now)

I am new to pg_routing, so just for my understanding, what would be the input to `pgr_createTopology`? Would that be an already existing `edge_table` (or a csv file with a source and target column?) through which the `edge_table` gets updated/created according to the `rows_where` parameter and the creation of a `vertex_table`? 

If so, that sounds like a feature I would definitely like to add to the list, seems quite useful and improves usability :)
Please let me know!

**Comment by tomkom** [comment]
Yes, an edge table that contains a geom column. Think of it as generating a graph def from a vector drawing. These are typically linestring edges (note, we distinguish vertices of a linestring, an graph nodes, where 2+ linestrings intersect). Duckdb implements spatial geometries as here https://duckdb.org/docs/extensions/spatial.html

Tolerance is for "snapping" linestrings if there is a small gap between end vertices.

**Comment by Dtenwolde** [comment]
I see, I believe there are opportunities to combine the spatial and duckpgq extension, so I'll keep this in mind. Thank you :)

**Comment by tomkom** [comment]
JUst catching up on this, did anyone manage to experiment with graph creation based on the DuckDB geometry support? This is now beautifully mainstream, and supporting graph based analytics ( shortest path) requires the ability to create a network from the geometries from a table that contains "lines". Is this something that is now tackled in the duckpg extension development?

**Comment by tomkom** [comment]
I see this got stale - unfortunate. Any chance this may be revived?

**Comment by tomkom** [comment]
Let me know if I can help somehow ( not sure my extension dev skill would be up to scratch, but can test). It would require a bit of spatial coding, which I can do (with the spatial extension - so maybe an unfortunate new dependency ( a heavy one). Maybe an external extension for it?

**Comment by Dtenwolde** [comment]
I don't want to add the spatial extension as a dependency, that's too much of a divergence from the goal of this extension. 

The first implementation of this pragma would basically execute the following query to make creating the vertex table easier:
`CREATE TABLE v AS  SELECT DISTINCT id FROM (SELECT src AS id FROM e UNION ALL SELECT dst AS id FROM e)`

I hope that also works for you as a first implementation.

**Comment by dbeatty10** [comment]
I've been playing around with using [`dbt-duckdb`](https://github.com/duckdb/dbt-duckdb) to manage data transformations expressed in SQL/PGQ.

Macros within dbt + Jinja can fill a similar role as functions in Postgres or pragmas in SQLite and DuckDB.

The ability to use Jinja templating and macros allows for expressing complex data transformation pipelines (like the logic within [`pgr_createTopology`](https://docs.pgrouting.org/latest/en/pgr_createTopology.html#id2)).

Disclaimer: I'm a maintainer for [`dbt-core`](https://github.com/dbt-labs/dbt-core) and an employee at [dbt Labs](https://www.getdbt.com/). Not trying to promote `dbt`, but it could come in handy for cases like this.

Two examples below (one without Jinja macros, and one with a Jinja macro).

### Initial setup

Both assume we start by creating the following table:

```sql
CREATE TABLE person_knows_person(person1id BIGINT, person2id BIGINT, createDate BIGINT);
INSERT INTO know VALUES (0,1, 10), (0,2, 11), (0,3, 12), (3,0, 13), (1,2, 14), (1,3, 15), (2,3, 16), (4,3, 17);
```

The "dbt-tonic" way to do this is with a CSV seed table like the following:

`seeds/person_knows_person.csv`

```csv
Person1Id,Person2Id,creationDate
0,1,10
0,2,11
0,3,12
3,0,13
1,2,14
1,3,15
2,3,16
4,3,17
```

Then create this table in the database:
```shell
dbt seed
```

### Creating the vertex table without a Jinja maco

`models/v4.sql`

```sql
{{ config(materialized="table") }}

SELECT DISTINCT id FROM (
    SELECT "Person1Id" AS id FROM {{ ref("person_knows_person") }}
    UNION ALL
    SELECT "Person2Id" AS id FROM {{ ref("person_knows_person") }}
)
```

### Creating the vertex table with a Jinja maco

`macros/create_vertex_table.sql`

```sql
{% macro create_vertex_table(edge_table, source_column, destination_column, id_column_name) %}

    SELECT DISTINCT {{ id_column_name }} FROM (
        SELECT {{ source_column }} AS {{ id_column_name }} FROM {{ edge_table }}
        UNION ALL
        SELECT {{ destination_column }} AS {{ id_column_name }} FROM {{ edge_table }}
    )

{% endmacro %}
```

`models/v4.sql`

```sql
{{ config(materialized="table") }}

{{ create_vertex_table(ref("person_knows_person"), "Person1Id", "Person2Id", "id") }}
```

### Commands and output

```shell
dbt build --select v4
dbt show --select v4 --quiet
```

```
| id |
| -- |
|  0 |
|  1 |
|  3 |
|  4 |
|  2 |
```

From this starting point, the other crucial data transformation steps within within [`pgr_createTopology`](https://docs.pgrouting.org/latest/en/pgr_createTopology.html#id2) could be added in subsequent transformation steps (expressed in dbt models). Alternatively, a [custom materialization](https://docs.getdbt.com/guides/create-new-materializations?step=1) in dbt could be created to handle these transformation steps in one shot.

**Comment by tomkom** [comment]
Thanks! The vertex table is easy, indeed as you have done. The edge table
(with possibly oriented arcs, based on analysis of topological
relationships) is harder (without the spatial ext.)
